flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/operations/utils/ValuesOperationFactory.java

```diff
@@ -0,0 +1,201 @@
+import org.apache.flink.table.expressions.ExpressionDefaultVisitor;
+import org.apache.flink.table.expressions.ResolvedExpression;
+import org.apache.flink.table.expressions.ValueLiteralExpression;
+import org.apache.flink.table.expressions.resolver.ExpressionResolver;
+import org.apache.flink.table.functions.BuiltInFunctionDefinitions;
+import org.apache.flink.table.operations.QueryOperation;
+import org.apache.flink.table.operations.ValuesQueryOperation;
+import org.apache.flink.table.types.DataType;
+import org.apache.flink.table.types.logical.LogicalType;
+import org.apache.flink.table.types.logical.utils.LogicalTypeGeneralization;
+import org.apache.flink.table.types.utils.TypeConversions;
+
+import javax.annotation.Nullable;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+import static org.apache.flink.table.expressions.ApiExpressionUtils.valueLiteral;
+
+/**
+ * Utility class for creating valid {@link ValuesQueryOperation} operation.
+ */
+@Internal
+class ValuesOperationFactory {
+	/**
+	 * Creates a valid {@link ValuesQueryOperation} operation.
+	 *
+	 * <p>It derives a row type based on {@link LogicalTypeGeneralization}. It flattens any
+	 * row constructors. It does not flatten ROWs which are a result of e.g. a function call.
+	 *
+	 * <p>The resulting schema can be provided manually. If it is not, the schema will be automatically derived from
+	 * the types of the expressions.
+	 */
+	QueryOperation create(
+			@Nullable TableSchema expectedSchema,
+			List<ResolvedExpression> resolvedExpressions,
+			ExpressionResolver.PostResolverFactory postResolverFactory) {
+		List<List<ResolvedExpression>> resolvedRows = unwrapFromRowConstructor(resolvedExpressions);
+
+		TableSchema schema = Optional.ofNullable(expectedSchema)
+			.orElseGet(() -> extractSchema(resolvedRows));
+
+		List<List<ResolvedExpression>> castedExpressions = resolvedRows.stream()
+			.map(row -> convertToExpectedRowType(postResolverFactory, schema.getFieldDataTypes(), row))
+			.collect(Collectors.toList());
+
+		return new ValuesQueryOperation(castedExpressions, schema);
+	}
+
+	private TableSchema extractSchema(List<List<ResolvedExpression>> resolvedRows) {
+		DataType[] dataTypes = findRowType(resolvedRows);
+		String[] fieldNames = IntStream.range(0, dataTypes.length)
+			.mapToObj(i -> "f" + i)
+			.toArray(String[]::new);
+		return TableSchema.builder()
+			.fields(fieldNames, dataTypes)
+			.build();
+	}
+
+	private List<ResolvedExpression> convertToExpectedRowType(
+			ExpressionResolver.PostResolverFactory postResolverFactory,
+			DataType[] dataTypes,
+			List<ResolvedExpression> row) {
+		return IntStream.range(0, row.size())
+			.mapToObj(i -> {
+				boolean typesMatch = row.get(i)
+					.getOutputDataType()
+					.getLogicalType()
+					.equals(dataTypes[i].getLogicalType());
```

Comment:
Use `LogicalTypeCast.supportsAvoidingCast` instead of `equals`.  This avoid unnecessary casts esp. rows with different field names.

Reply:
In most of the cases `LogicalType#equals` is too specific.